### PR TITLE
clustermesh: Make IPCache CPlane aware of the ClusterID

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
This is a part of the Cluster Mesh with overlapping PodCIDR support. This PR extends IPCache CPlane (specifically, IPIdentityWatcher) to take ClusterID. Please see the commit messages for more details. Note that this PR only adds infrastructure for taking ClusterID and no one actually uses it yet.

```release-note
clustermesh: Make IPCache CPlane aware of the ClusterID
```

### CODE REVIEW

**Review by gandro — COMMENTED** [review]
Thanks for the PR. I have yet to do an in-depth review, but if I understand this correctly, you are replacing the IPCache key (which currently is a prefix) with prefix+clusterid.

The PR seems to address the legacy Upsert/Delete logic, I think it does not change how the newer asynchronous `UpsertLabels` infrastructure behaves. Is that intended? 

Ultimately, we will remove `Upsert` and `Delete` as part of https://github.com/cilium/cilium/issues/21142 and only have "UpsertLabels/RemoveLabels" instead - so I wonder if you have plans to address that too?

**Comment by gandro** [comment]
Also cc @joestringer @christarazi - since this affects the public interface of IPCache, I'd like to get a second pair of eyes on this.

File: pkg/clustermesh/types/addressing.go
Comment (christarazi) [inline pkg/clustermesh/types/addressing.go:241]: It'd be good to add unit tests for this.

File: pkg/ipcache/ipcache.go
Comment (christarazi) [inline pkg/ipcache/ipcache.go:369]: Now that we are adding two new formats for `ip string`, I think we should document at the top of this function what the different formats that `ip` can be.

File: pkg/ipcache/kvstore.go
Comment (christarazi) [inline pkg/ipcache/kvstore.go]: Nit: 

```
return NewClusterIPIdentityWatcher(0, ipc, backend)
```

**Review by christarazi — CHANGES_REQUESTED** [review]
Changes look mostly good to me, I have a few comments to address. Biggest concern is the lack of unit testing for the parsing of the prefix cluster type and additional documentation on the ipcache upsert.

@gandro Regarding the new IPcache API, firstly I don't see any calls to the old API. I only see that this PR is changing the underlying upsert logic to handle the clustermesh-specific format. This clustermesh-specific case might actually be one of the use cases for `UpsertIdentity` from the new API, because I believe we already know the identity at this point. Overall, I think this PR can still be compatible with the new API, unless I've missed something.

**Comment by YutaroHayakawa** [comment]
Fixed the points Chris pointed out. Thanks!

**Comment by gandro** [comment]
> @gandro Regarding the new IPcache API, firstly I don't see any calls to the old API. I only see that this PR is changing the underlying upsert logic to handle the clustermesh-specific format.

You are right, I misread the code. It's only affecting private `upsertLocked`, not the legacy `Upsert` as I originally thought. Agreed then that this should not cause issues.

**Review by gandro — APPROVED** [review]

**Review by christarazi — APPROVED** [review]

**Comment by YutaroHayakawa** [comment]
Oops, some tests are crashing with ipcache sync... Let me check.

**Comment by YutaroHayakawa** [comment]
Ok, this was the cause. I overwrite `cidrCluster` variable with new one.

```go
	var cidrCluster cmtypes.PrefixCluster <==== Here
	cacheModification := Delete
	oldHostIP, encryptKey := ipc.getHostIPCache(ip)
	oldK8sMeta := ipc.getK8sMetadata(ip)
	var newHostIP net.IP
	var oldIdentity *Identity
	newIdentity := cachedIdentity
	callbackListeners := true

	var err error
	if cidrCluster, err = cmtypes.ParsePrefixCluster(ip); err == nil {
		// Check whether the deleted CIDR was shadowed by an endpoint IP. In
		// this case, skip calling back the listeners since they don't know
		// about its mapping.
		if _, endpointIPFound := ipc.ipToIdentityCache[cidrCluster.AddrCluster().String()]; endpointIPFound {
			scopedLog.Debug("Deleting CIDR shadowed by endpoint IP")
			callbackListeners = false
		}
	} else if addrCluster, err := cmtypes.ParseAddrCluster(ip); err == nil { // Endpoint IP or Endpoint IP with ClusterID
		// Convert the endpoint IP into an equivalent full CIDR.
		cidrCluster := addrCluster.AsPrefixCluster() <==== Overwritten
```

**Comment by YutaroHayakawa** [comment]
k8s-1.26-kernel-net-next: Known flake https://github.com/cilium/cilium/issues/22579
runtime: Known flake https://github.com/cilium/cilium/issues/23176
ConformanceIngress: New flake https://github.com/cilium/cilium/issues/23179

**Comment by YutaroHayakawa** [comment]
/test-1.26-net-next

**Comment by YutaroHayakawa** [comment]
Oh, it looks like the ConformanceIngress is failing consistently. The reason is I am not in sync with the latest master and don't have in revert commit. Once all other tests are succeeded, I'll take a record and rebase the branch.

https://github.com/cilium/cilium/pull/23169

**Comment by YutaroHayakawa** [comment]
Hmm, runtime test is still failing, looks like a known flake, but that could be a regression. Let me investigate.

**Comment by YutaroHayakawa** [comment]
Looks like this is another flake of the new LB IPAM feature. I filed a new issue.

https://github.com/cilium/cilium/issues/23185

**Comment by YutaroHayakawa** [comment]
Ok, runtime test passed. I'll record the current state and rebase this branch to apply https://github.com/cilium/cilium/pull/23169.

スクリーンショット 2023-01-20 0 18 06

**Comment by YutaroHayakawa** [comment]
The upstream SBOM image build workflow had an error. I'll rerun all workflows.

**Comment by YutaroHayakawa** [comment]
No, the issue was still there

**Review by jrajahalme — CHANGES_REQUESTED** [review]
I see the dataplane changes are not in yet, but when you get to push the new `PrefixCluster` types to the (LPM) bpf map, you'll find out that the prefix part of the key needs to be at the end of the key for the longest-prefix match to work. Given this, the cluster ID in the key needs to be before the prefix, and then the new type would be better named as `ClusterPrefix`.

**Comment by YutaroHayakawa** [comment]
> I see the dataplane changes are not in yet, but when you get to push the new PrefixCluster types to the (LPM) bpf map, you'll find out that the prefix part of the key needs to be at the end of the key for the longest-prefix match to work. Given this, the cluster ID in the key needs to be before the prefix

I already changed the IPCache map in [this commit](https://github.com/cilium/cilium/pull/22200/commits/723b77542f1a7e9210eb4bf0f5dccb56484219fa#diff-60e54b50afef4f14bf4ab0718f98dd3c73b1067143ef89e1fb0c60a51a85c8a9R171). As you said, I placed the ClusterID part before prefix.

> and then the new type would be better named as ClusterPrefix

Do you mean because the order in the map is ClusterID -> Prefix? Hmm, in that sense, tunnel maps also use this type and their order is prefix -> ClusterID. So, I think the order in the map can be a separated topic. Also, we already have a `AddrCluster` type, so from a consistency standpoint, I want to go with `PrefixCluster`.

**Comment by YutaroHayakawa** [comment]
I pulled the latest master to get the CI fix. Let me run tests again.

**Comment by YutaroHayakawa** [comment]
- runtime: A new flake. Filed an issue. https://github.com/cilium/cilium/issues/23285

**Comment by YutaroHayakawa** [comment]
Rebased on the latest master and fixed conflicts.

**Comment by YutaroHayakawa** [comment]
Travis: Known flake and succeeding on ARM64 https://github.com/cilium/cilium/issues/17228

Let me make this ready-to-merge. The naming stuff pointed out by Jarno can be fixed later if we think it's and issue.

**Comment by qmonnet** [comment]
@ldelossa Can we get a quick review from @cilium/cli as codeowner please?

**Comment by pchaigno** [comment]
CLI changes are very minor, I think it's fine to skip the review for that. Merging.

**Pull request merged**
